biosys/apps/main/models.py

```diff
@@ -68,60 +70,141 @@ def schema_model(self):
+            schema = resource.get('schema', {})
             try:
                 # use frictionless validator
                 jsontableschema.validate(schema)
-            except Exception as e:
+            except Exception:
                 raise ValidationError(
-                    'Schema errors for resource "{}": {}'.format(
-                        self.resource.get('name'),
-                        [e.message for e in jsontableschema.validator.iter_errors(schema)]))
+                    'Schema errors for resource "{}":<br>{}'.format(
+                        resource.get('name'),
+                        "<br>".join([e.message for e in jsontableschema.validator.iter_errors(schema)])
+                    ))
             try:
                 # use our own schema class to validate.
                 # The constructor should raise an exception if error
-                if self.type == self.TYPE_SPECIES_OBSERVATION:
+                if dataset_type == Dataset.TYPE_SPECIES_OBSERVATION:
                     SpeciesObservationSchema(schema)
-                elif self.type == self.TYPE_OBSERVATION:
+                elif dataset_type == Dataset.TYPE_OBSERVATION:
                     ObservationSchema(schema)
                 else:
                     GenericSchema(schema)
             except Exception as e:
                 raise ValidationError(
                     'Schema errors for resource "{}": {}'.format(
-                        self.resource.get('name'),
+                        resource.get('name'),
                         e))
 
+    def validate_data(self, data):
+        if not data:
+            msg = "field 'data' cannot be null or empty"
+            raise ValidationError(msg)
+        try:
+            field_errors = self.schema.get_error_fields(data)
+            if len(field_errors) > 0:
+                for field_name, data in field_errors:
+                    msg = "'{}': {}".format(field_name, data.get('error'))
+                    raise ValidationError(msg)
+        except Exception as e:
+            raise ValidationError(e)
+
+    def clean(self):
+        """
+        Validate the data descriptor
+        """
+        # Validate the data package
+        self.validate_data_package(self.data_package, self.type)
+
+    def is_custodian(self, user):
+        return self.project.is_custodian(user)
+
+    # API permissions
+    @staticmethod
+    def has_read_permission(request):
+        return True
+
+    def has_object_read_permission(self, request):
+        return True
+
+    @staticmethod
+    def has_create_permission(request):
+        """
+        Custodian and admin only
+        Check that the user is a custodian of the project pk passed in the POST data
+        :param request:
+        :return:
+        """
+        result = False
+        if is_admin(request.user):
+            result = True
+        else:
+            try:
+                project = Project.objects.get(pk=request.data['project'])
```

Comment:
:-) I don't know what kind of evil got me.
Why I don't specify the exception is that I want also to catch the error if 'project' in not in the `request.data`. It's laziness.